dockstore-webservice/src/main/java/io/dockstore/webservice/core/AuthenticatedUser.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright 2023 OICR and UCSC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.dockstore.webservice.core;
+
+import java.util.Optional;
+
+/**
+ * Has the authenticated user, if any, of the current request to the API.
+ */
+public final class AuthenticatedUser {
+    private static final ThreadLocal<User> AUTH_USER = new ThreadLocal<>();
```

Comment:
There's a SonarCloud warning about there not being a `remove`, but if I invoked `remove` in a servlet response filter, it broke things. We set the ThreadLocal at the start of every request, so it should be good.

Reply:
In this implementtion, the `ThreadLocal` value that's set will persist until the next time the thread is used to handle a request, causing the `User` and whatever it references to be retained.  It'd be good to call `remove` (or `set(null)`) at the end of the request, to prevent the formentioned problem, and to help ensure that we don't accidentally leak a `User` between requests.  According to the Internets, the ThreadLocal implementation has been improved, so that dangling ThreadLocals disappear when a thread terminates, so we don't need to worry about that anymore.

Reply:
I tried again and debugged, and unfortunately the servlet filter is invoked before the the response is JSONified, so calling remove doesn't work. I'm a little uncomfortable with this, but I think we have to go with it.

Reply:
Yeah, it's probably ok (I hope).  Apparently, Jersey supports an injectable `CloseableService` that you can register objects with that should be closed after the request, see the first answer here:
https://stackoverflow.com/questions/53433174/jersey-close-i-o-resources-after-http-response
If we created a `Closeable` which would `remove()` the `ThreadLocal` when it was closed, and could automatically register it with the underlying `CloseableService` at the beginning of each request, that would do it, I think.  Not sure if that's possible, though.

Reply:
To summarize:

1. The nice thing about the current approach is that it works across the board, for all endpoints that stream `User`, even when they're nested objects. The potential downside, is that if a thread gets reused without going through the ServletFilters, and the AuthUser ends up being set  to the wrong user. I haven't seen this, and every new endpoint request does go through the filter, so those **should** always be good, but in theory there might be some other case where it might cause a problem.
2. We could go back to the original approach, which is that specific endpoints, probably `/users/user` and `/users/username/{username}`  be the only endpoints that return the real values of admin and curator. If we did that, I wouldn't use a custom object mapper nor a servlet filter, but instead use the Closeable you suggested above, injecting the AuthUser in the specific request(s).

Any thoughts/preferences?

Reply:
> Any thoughts/preferences?

I like the current approach (number 1), although it's upsetting to not able to clean up at the end of the request. 
 I felt like there had to be a way to configure Dropwizard to run some code in the handler thread before/after each request/response, but couldn't find one.  Fortunately, if the app server messes up and a `User` crosses over into an alien thread, the worst case is that the admin/curator statuses are displayed when they're not supposed to, a pretty minor blip in the spectrum of oopsies...